samanthas_telegram_bot/api_clients/base/base_api_client.py

```diff
@@ -0,0 +1,274 @@
+from samanthas_telegram_bot.api_clients.auxil.models import NotificationParamsForStatusCode
+from samanthas_telegram_bot.api_clients.base.exceptions import BaseApiClientError
+from samanthas_telegram_bot.auxil.log_and_notify import logs
+from samanthas_telegram_bot.data_structures.constants import CALLER_LOGGING_STACK_LEVEL
+from samanthas_telegram_bot.data_structures.context_types import CUSTOM_CONTEXT_TYPES
+
+logger = logging.getLogger(__name__)
+
+
+class BaseApiClient:
+    @classmethod
+    async def get(
+        cls,
+        update: Update,
+        context: CUSTOM_CONTEXT_TYPES,
+        url: str,
+        notification_params_for_status_code: NotificationParamsForStatusCode,
+        headers: dict[str, str] | None = None,
+        params: DataDict | None = None,
+    ) -> tuple[int, DataDict]:
+        """Makes a GET request, returns a tuple containing status code and JSON data.
+
+        Note:
+            Only pass **informative** status codes in ``notification_params_for_status_code``.
+            E.g. If ``404 NOT FOUND`` means something relevant, pass it in, along with
+            notification parameters (how to log and notify admins). For all other cases,
+            let the API client raise its own exception and handle it accordingly
+            (e.g. with an exception handler in the bot).
+        """
+        return await cls._make_request_and_get_data(
+            method=HttpMethod.GET,
+            update=update,
+            context=context,
+            url=url,
+            headers=headers,
+            params=params,
+            notification_params_for_status_code=notification_params_for_status_code,
+        )
+
+    @classmethod
+    async def post(
+        cls,
+        update: Update,
+        context: CUSTOM_CONTEXT_TYPES,
+        url: str,
+        notification_params_for_status_code: NotificationParamsForStatusCode,
+        headers: dict[str, str] | None = None,
+        data: DataDict | None = None,
+        json_data: DataDict | None = None,
+        params: DataDict | None = None,
+    ) -> tuple[int, DataDict]:
+        """Makes a POST request, returns a tuple containing status code and JSON data.
+
+        The ``data`` parameter is passed as ``data``, ``json_data`` as ``json`` to
+        httpx client. You can pass either one of them, but not both at the same time.
+
+        Note:
+            Only pass **informative** status codes in ``notification_params_for_status_code``.
+            E.g. If ``404 NOT FOUND`` means something relevant, pass it in, along with
+            notification parameters (how to log and notify admins). For all other cases,
+            let the API client raise its own exception and handle it accordingly
+            (e.g. with an exception handler in the bot).
+        """
+        if data is None and json_data is None:
+            raise TypeError("Either `data` or `json_data` must be provided. You passed nothing.")
+
+        if not (data is None or json_data is None):
```

Comment:
Imho

`if data is not None and json_data is not None` 
is more readable

Reply:
I know, that's what I put in the beginning and then remembered some clever book that advised to replace that with parenthesized expressions :) 

I changed it back, though. It is more readable, of course.